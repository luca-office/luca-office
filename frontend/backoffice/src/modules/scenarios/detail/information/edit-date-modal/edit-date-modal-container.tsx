import * as React from "react"
import {useScenario, useUpdateScenario} from "shared/graphql/hooks"
import {Scenario} from "shared/models"
import {EditScenarioDateModal} from "./edit-date-modal"

export interface Props {
  readonly scenario: Scenario
  readonly onDismiss: () => void
  readonly onConfirm: () => void
}

export const EditScenarioDateModalContainer: React.FC<Props> = ({onConfirm, onDismiss, scenario}) => {
  const {scenario: scenarioOption} = useScenario(scenario.id)
  const {updateScenario} = useUpdateScenario()

  const [isFictiveDateSelected, setIsFictiveDateSelected] = React.useState(scenario.date !== null)
  const [selectedFictiveDate, setSelectedFictiveDate] = React.useState<Date>(
    scenario.date !== null ? new Date(scenario.date) : new Date()
  )

  const handleSubmit = (date: string | null | undefined) => {
    scenarioOption.forEach(scenario => {
      updateScenario(scenario.id, {
        description: scenario.description,
        name: scenario.name,
        shouldDisplayTime: scenario.shouldDisplayTime,
        shouldHideReferenceBookChapters: scenario.shouldHideReferenceBookChapters,
        tags: scenario.tags,
        completionEmailAddress: scenario.completionEmailAddress,
        date,
        introductionEmailId: scenario.introductionEmailId,
        sampleCompanyId: scenario.sampleCompanyId,
        maxDurationInSeconds: scenario.maxDurationInSeconds
      }).then(onConfirm)
    })
  }

  return (
    <EditScenarioDateModal
      setSelectedFictiveDate={setSelectedFictiveDate}
      selectedFictiveDate={selectedFictiveDate}
      isFictiveDateSelected={isFictiveDateSelected}
      setIsFictiveDateSelected={setIsFictiveDateSelected}
      onConfirm={handleSubmit}
      onDismiss={onDismiss}
    />
  )
}
