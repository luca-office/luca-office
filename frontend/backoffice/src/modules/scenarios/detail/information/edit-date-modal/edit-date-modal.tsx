import {css} from "@emotion/react"
import {isArray} from "lodash-es"
import * as React from "react"
import {Modal, Overlay, SelectableCard, SelectionIconType, SingleDatePicker, Text} from "shared/components"
import {Flex, spacingCard, spacingLarge, spacingMedium} from "shared/styles"
import {useLucaTranslation} from "shared/translations"

export interface EditDateModalProps {
  readonly isFictiveDateSelected: boolean
  readonly onConfirm: (date: string | null | undefined) => void
  readonly onDismiss: () => void
  readonly selectedFictiveDate: Date
  readonly setIsFictiveDateSelected: React.Dispatch<React.SetStateAction<boolean>>
  readonly setSelectedFictiveDate: React.Dispatch<React.SetStateAction<Date>>
}

export const EditScenarioDateModal: React.FC<EditDateModalProps> = ({
  onDismiss,
  selectedFictiveDate,
  isFictiveDateSelected,
  setIsFictiveDateSelected,
  setSelectedFictiveDate,
  onConfirm
}) => {
  const {t} = useLucaTranslation()

  return (
    <Overlay>
      <Modal
        customStyles={styles.modal}
        title={t("scenario_details__date_edit")}
        onDismiss={onDismiss}
        onConfirm={() => onConfirm(isFictiveDateSelected ? selectedFictiveDate.toISOString() : null)}
        confirmButtonDisabled={isFictiveDateSelected && !selectedFictiveDate}>
        <>
          <Text customStyles={styles.instruction}>{t("scenario_details__date_edit_modal_text")}</Text>
          <SelectableCard
            customStyles={styles.actualDateCard}
            selectionIconType={SelectionIconType.RADIO}
            onClick={() => setIsFictiveDateSelected(false)}
            selected={!isFictiveDateSelected}
            title={t("scenario_details__actual_date")}
            text={t("scenario_details__date_edit_modal_actual_date")}
          />
          <SelectableCard
            selectionIconType={SelectionIconType.RADIO}
            selected={isFictiveDateSelected}
            customStyles={styles.fictiveDateCard}
            onClick={() => setIsFictiveDateSelected(true)}
            title={t("scenario_details__fictive_date")}
            text={t("scenario_details__date_edit_modal_fictive_date")}
            customContent={
              <SingleDatePicker
                customStyles={Flex.column}
                customDatePickerStyles={styles.datePicker}
                onChange={(date: Date | Date[]) => setSelectedFictiveDate(isArray(date) ? date[0] : date)} // is never an array -> avoid type errors
                value={selectedFictiveDate}
              />
            }
          />
        </>
      </Modal>
    </Overlay>
  )
}

const styles = {
  modal: css({
    width: "40vw",

    header: {
      marginBottom: spacingCard
    }
  }),
  instruction: css({
    marginBottom: spacingMedium
  }),
  actualDateCard: css({
    marginBottom: spacingMedium
  }),
  fictiveDateCard: css({
    marginBottom: spacingLarge
  }),
  datePicker: css({
    margin: spacingMedium
  })
}
