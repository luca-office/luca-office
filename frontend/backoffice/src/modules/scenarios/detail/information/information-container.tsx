import * as React from "react"
import {ScenarioUpdate} from "shared/graphql/generated/globalTypes"
import {Scenario} from "shared/models"
import {Information} from "./information"

export interface InformationContainerProps {
  readonly duplicateScenarioLoading: boolean
  readonly isFinalizeScenarioLoading: boolean
  readonly isPublishScenarioLoading: boolean
  readonly scenario: Scenario
  readonly scenarioContributorsCount: number
  readonly updateInProgress: boolean
  readonly updateScenario: (id: string, update: ScenarioUpdate) => void
  readonly readonly?: boolean
}

export const InformationContainer: React.FC<InformationContainerProps> = props => {
  const [isInviteContributorsModalVisible, setIsInviteContributorsModalVisible] = React.useState(false)
  const [isEditDateModalVisible, setIsEditDateModalVisible] = React.useState(false)
  const [isEditDurationModalVisible, setIsEditDurationModalVisible] = React.useState(false)

  const toggleIsInviteContributorsModalVisible = () =>
    setIsInviteContributorsModalVisible(!isInviteContributorsModalVisible)

  const toggleIsEditDateModalVisible = () => setIsEditDateModalVisible(!isEditDateModalVisible)

  const toggleIsEditDurationModalVisible = () => setIsEditDurationModalVisible(!isEditDurationModalVisible)

  return (
    <Information
      isEditDateModalVisible={isEditDateModalVisible}
      isInviteContributorsModalVisible={isInviteContributorsModalVisible}
      isEditDurationModalVisible={isEditDurationModalVisible}
      toggleIsEditDateModalVisible={toggleIsEditDateModalVisible}
      toggleIsInviteContributorsModalVisible={toggleIsInviteContributorsModalVisible}
      toggleIsEditDurationModalVisible={toggleIsEditDurationModalVisible}
      {...props}
    />
  )
}
