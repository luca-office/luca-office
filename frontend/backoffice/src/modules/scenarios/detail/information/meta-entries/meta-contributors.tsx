import {css} from "@emotion/react"
import * as React from "react"
import {Icon, ReadonlyActionField, Text} from "shared/components"
import {IconName} from "shared/enums"
import {Flex, spacingSmall, TextSize} from "shared/styles"
import {useLucaTranslation} from "shared/translations"

export interface MetaContributorsProps {
  readonly scenarioContributorsCount: number
  readonly onClick: () => void
  readonly disabled?: boolean
}

export const MetaContributors: React.FC<MetaContributorsProps> = ({scenarioContributorsCount, onClick, disabled}) => {
  const {t} = useLucaTranslation()
  return (
    <>
      <ReadonlyActionField
        label={t("scenario_details__general_contributors")}
        buttonLabel={t("scenario_details__button_invite")}
        disabled={disabled}
        onClick={onClick}
        renderValue={() => (
          <div css={Flex.row}>
            <Icon name={IconName.User} />
            <Text customStyles={styles.smallPaddingLeft} size={TextSize.Medium}>
              {`${scenarioContributorsCount} ${t("scenario_details__general_contributors")}`}
            </Text>
          </div>
        )}
      />
    </>
  )
}

const styles = {
  statusLabel: css({
    marginLeft: spacingSmall
  }),
  smallPaddingLeft: css({
    paddingLeft: spacingSmall
  })
}
