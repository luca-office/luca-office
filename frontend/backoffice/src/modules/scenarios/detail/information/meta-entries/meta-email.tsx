import {css} from "@emotion/react"
import * as React from "react"
import {useDispatch} from "react-redux"
import {Icon, ReadonlyActionField, Text} from "shared/components"
import {IconName} from "shared/enums"
import {EmailDirectory} from "shared/graphql/generated/globalTypes"
import {Scenario} from "shared/models"
import {Flex, spacingSmall, TextSize} from "shared/styles"
import {useLucaTranslation} from "shared/translations"
import {navigateToRouteAction} from "../../../../../redux/actions/navigation-action"
import {Route} from "../../../../../routes"

export interface MetaEmailProps {
  readonly scenario: Scenario
  readonly isEditable: boolean
}

export const MetaEmail: React.FC<MetaEmailProps> = ({scenario, isEditable}) => {
  const dispatch = useDispatch()

  const {t} = useLucaTranslation()

  const hasIntroductionEmail = !!scenario.introductionEmailId

  const showCreateModal = () =>
    dispatch(navigateToRouteAction(Route.ScenarioDetailIntroductionEmailCreation, {scenarioId: scenario.id}))
  const navigateToEmail = (emailId: UUID) =>
    dispatch(
      navigateToRouteAction(Route.ScenarioEmails, {
        scenarioId: scenario.id,
        emailId,
        directory: EmailDirectory.Inbox
      })
    )
  const handleClick = () => (hasIntroductionEmail ? navigateToEmail(scenario.introductionEmailId!) : showCreateModal())

  return (
    <ReadonlyActionField
      label={t("scenario_details__general_email")}
      buttonLabel={!hasIntroductionEmail ? t("create_button") : isEditable ? t("edit_button") : t("show_button")}
      disabled={!isEditable && !hasIntroductionEmail}
      onClick={handleClick}
      renderValue={() => (
        <div css={Flex.row}>
          <Icon name={IconName.Email} />
          <Text customStyles={styles.smallPaddingLeft} size={TextSize.Medium}>
            {t(hasIntroductionEmail ? "scenario_details__available" : "scenario_details__unavailable")}
          </Text>
        </div>
      )}
    />
  )
}

const styles = {
  statusLabel: css({
    marginLeft: spacingSmall
  }),
  smallMarginRight: css({
    marginRight: spacingSmall
  }),
  smallPaddingLeft: css({
    paddingLeft: spacingSmall
  })
}
