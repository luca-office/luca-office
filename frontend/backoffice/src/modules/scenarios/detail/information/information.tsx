import * as React from "react"
import {CardContent, Column, Columns, Icon, Label, ReadonlyActionField, Text} from "shared/components"
import {IconName} from "shared/enums"
import {ProjectUpdate, ScenarioUpdate} from "shared/graphql/generated/globalTypes"
import {Flex, TextSize} from "shared/styles"
import {useLucaTranslation} from "shared/translations"
import {convertSecondsToMinutes, formatDate} from "shared/utils"
import {
  InlineEditableHeaderContainer,
  InlineEditableTextareaContainer,
  MetaEntryTags,
  OverlayEditCompositeFieldConfig,
  OverlayEditField,
  OverlayEditFieldType
} from "../../../../components"
import {EditDurationModalContainer} from "../../../../components/edit-duration-modal/edit-duration-modal-container"
import {InviteScenarioContributorsModalContainer} from "../invite-contributors/scenario-invite-contributors-container"
import {EditScenarioDateModalContainer} from "./edit-date-modal/edit-date-modal-container"
import {informationStyles as styles} from "./information.style"
import {InformationContainerProps} from "./information-container"
import {MetaContributors} from "./meta-entries/meta-contributors"
import {MetaEmail} from "./meta-entries/meta-email"

export interface InformationProps extends InformationContainerProps {
  readonly toggleIsEditDateModalVisible: () => void
  readonly toggleIsEditDurationModalVisible: () => void
  readonly toggleIsInviteContributorsModalVisible: () => void
  readonly isInviteContributorsModalVisible: boolean
  readonly isEditDateModalVisible: boolean
  readonly isEditDurationModalVisible: boolean
}

export const Information: React.FC<InformationProps> = ({
  scenario,
  updateInProgress,
  isFinalizeScenarioLoading,
  isPublishScenarioLoading,
  isInviteContributorsModalVisible,
  isEditDateModalVisible,
  isEditDurationModalVisible,
  duplicateScenarioLoading,
  scenarioContributorsCount,
  updateScenario,
  readonly,
  toggleIsInviteContributorsModalVisible,
  toggleIsEditDateModalVisible,
  toggleIsEditDurationModalVisible
}) => {
  const {t} = useLucaTranslation()
  const isEditable = !readonly
  const fieldIcon = !isEditable ? IconName.LockClosed : IconName.EditPencil
  const canHandleUpdate =
    isEditable &&
    !updateInProgress &&
    !isFinalizeScenarioLoading &&
    !isPublishScenarioLoading &&
    !duplicateScenarioLoading

  const descriptionField: OverlayEditCompositeFieldConfig = {
    updateId: "description",
    value: scenario.description,
    labelKey: "description",
    type: OverlayEditFieldType.TEXTAREA
  }

  const handleUpdate = (update: Partial<ScenarioUpdate>) => {
    if (!canHandleUpdate) {
      return
    }
    updateScenario(scenario.id, {
      name: update.name ? update.name : scenario.name,
      description: update.description ? update.description : scenario.description,
      tags: update.tags ? update.tags : scenario.tags,
      maxDurationInSeconds: scenario.maxDurationInSeconds,
      shouldDisplayTime: true, // this is not implemented in UI in MVP,
      shouldHideReferenceBookChapters: scenario.shouldHideReferenceBookChapters,
      introductionEmailId: scenario.introductionEmailId,
      completionEmailAddress: scenario.completionEmailAddress,
      sampleCompanyId: scenario.sampleCompanyId,
      date: scenario.date
    })
  }

  return (
    <React.Fragment>
      <CardContent>
        <Columns customStyles={styles.titleAndTime}>
          <Column flexGrow={3}>
            <Label label={t("title")} icon={fieldIcon} />
            <InlineEditableHeaderContainer
              onConfirm={name => handleUpdate({name})}
              text={scenario.name}
              customStyles={styles.detailHeaderName}
              disabled={!isEditable}
            />
          </Column>
          <Column>
            <ReadonlyActionField
              buttonLabel={t("edit_button")}
              customStyles={styles.detailHeaderTime}
              onClick={isEditable ? toggleIsEditDurationModalVisible : undefined}
              label={t("projects__detail_time_label")}
              disabled={!isEditable}
              renderValue={() => (
                <Text size={TextSize.Medium} customStyles={styles.noTime}>
                  <Icon name={IconName.Clock} customStyles={styles.noTimeIcon} />
                  {scenario.maxDurationInSeconds === null
                    ? t("placeholder__no_entry")
                    : `${convertSecondsToMinutes(scenario.maxDurationInSeconds)} ${t("unit__minutes_short")}`}
                </Text>
              )}
            />
          </Column>
        </Columns>

        <React.Fragment>
          <Column customStyles={styles.descriptionColumn}>
            <Label label={t("description")} icon={fieldIcon} />
            <OverlayEditField<ProjectUpdate>
              formFields={[descriptionField]}
              fieldLabelKey={"description"}
              dialogTitleKey={"scenario__overlay_update_description"}
              onUpdate={handleUpdate}
              updateLoading={
                updateInProgress || isFinalizeScenarioLoading || isPublishScenarioLoading || duplicateScenarioLoading
              }
              disabled={!isEditable}
              renderValue={() => (
                <InlineEditableTextareaContainer text={scenario.description} readOnly={true} disabled={!isEditable} />
              )}
              displayPlain={true}
              customStyles={styles.description}
            />
          </Column>
          <div css={[Flex.column, styles.metadataWrapper]}>
            <MetaEntryTags
              tags={scenario.tags}
              updateLoading={updateInProgress}
              handleUpdate={tags => handleUpdate({tags})}
              disabled={!isEditable}
            />

            <Columns customStyles={styles.metadataRowMarginTop}>
              <Column>
                <MetaEmail isEditable={isEditable} scenario={scenario} />
              </Column>
              <Column>
                <ReadonlyActionField
                  label={t("scenario_details__date")}
                  buttonLabel={isEditable ? t("edit_button") : undefined}
                  onClick={isEditable ? toggleIsEditDateModalVisible : undefined}
                  renderValue={() => (
                    <div css={Flex.row}>
                      <Icon name={IconName.Calendar} />
                      <Text customStyles={styles.smallPaddingLeft} size={TextSize.Medium}>
                        {scenario.date !== null
                          ? `${formatDate(new Date(scenario.date))} (${t("scenario_details__fictive_date")})`
                          : `${t("scenario_details__actual_date")}`}
                      </Text>
                    </div>
                  )}
                />
              </Column>
              <Column customStyles={styles.smallPaddingLeft}>
                <MetaContributors
                  disabled={!isEditable}
                  onClick={toggleIsInviteContributorsModalVisible}
                  scenarioContributorsCount={scenarioContributorsCount}
                />
              </Column>
            </Columns>
          </div>
        </React.Fragment>
      </CardContent>
      {isInviteContributorsModalVisible && (
        <InviteScenarioContributorsModalContainer
          scenarioId={scenario.id}
          onDismiss={toggleIsInviteContributorsModalVisible}
        />
      )}
      {isEditDateModalVisible && (
        <EditScenarioDateModalContainer
          scenario={scenario}
          onConfirm={toggleIsEditDateModalVisible}
          onDismiss={toggleIsEditDateModalVisible}
        />
      )}
      {isEditDurationModalVisible && (
        <EditDurationModalContainer
          onDismiss={toggleIsEditDurationModalVisible}
          onConfirm={toggleIsEditDurationModalVisible}
          scenarioId={scenario.id}
          scenarioDurationInSeconds={scenario.maxDurationInSeconds ?? 0}
        />
      )}
    </React.Fragment>
  )
}
