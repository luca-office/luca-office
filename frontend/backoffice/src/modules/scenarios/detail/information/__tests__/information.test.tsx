import {shallow} from "enzyme"
import * as React from "react"
import {create} from "react-test-renderer"
import {Column, Columns, ReadonlyActionField} from "shared/components"
import {scenariosMock} from "shared/graphql/__mocks__"
import {ScenarioUpdate} from "shared/graphql/generated/globalTypes"
import {Option} from "shared/utils"
import {getDateValueMock} from "sharedTests/utils/date-mock"
import {InlineEditableHeaderContainer, MetaEntryTags, OverlayEditField} from "../../../../../components"
import {EditDurationModalContainer} from "../../../../../components/edit-duration-modal/edit-duration-modal-container"
import * as useScenarioUpdateHook from "../../hooks/use-scenario-update"
import {UseScenarioUpdateHook} from "../../hooks/use-scenario-update"
import {InviteScenarioContributorsModalContainer} from "../../invite-contributors/scenario-invite-contributors-container"
import {EditScenarioDateModalContainer} from "../edit-date-modal/edit-date-modal-container"
import {Information, InformationProps} from "../information"
import {MetaContributors} from "../meta-entries/meta-contributors"
import {MetaEmail} from "../meta-entries/meta-email"

const scenario = scenariosMock[0]
const scenarioId = scenario.id

const defaultUpdateData: ScenarioUpdate = {
  name: scenario.name,
  description: scenario.description,
  tags: scenario.tags,
  maxDurationInSeconds: scenario.maxDurationInSeconds,
  shouldDisplayTime: true, // this is not implemented in UI in MVP,
  shouldHideReferenceBookChapters: scenario.shouldHideReferenceBookChapters,
  introductionEmailId: scenario.introductionEmailId,
  completionEmailAddress: scenario.completionEmailAddress,
  sampleCompanyId: scenario.sampleCompanyId,
  date: scenario.date
}

const defaultProps: InformationProps = {
  scenario: scenariosMock[0],
  updateInProgress: false,
  isInviteContributorsModalVisible: false,
  scenarioContributorsCount: 0,
  toggleIsInviteContributorsModalVisible: jest.fn(),
  toggleIsEditDateModalVisible: jest.fn(),
  toggleIsEditDurationModalVisible: jest.fn(),
  isFinalizeScenarioLoading: false,
  isPublishScenarioLoading: false,
  isEditDateModalVisible: false,
  isEditDurationModalVisible: false,
  duplicateScenarioLoading: false,
  updateScenario: jest.fn()
}

const scenarioUpdateHookValuesDefault: UseScenarioUpdateHook = {
  updateScenario: jest.fn(() => Promise.resolve(Option.of(scenario))),
  updateScenarioLoading: false
}

const scenarioUpdateStateSpy = jest.spyOn(useScenarioUpdateHook, "useScenarioUpdate")

const getComponent = (props?: Partial<InformationProps>) => <Information {...{...defaultProps, ...props}} />

describe("information", () => {
  const _Date = Date
  beforeEach(() => {
    const mockDate = getDateValueMock(1970, 1, 1)
    Date = mockDate
  })
  afterEach(() => {
    Date = _Date
  })

  it("renders correctly", async () => {
    scenarioUpdateStateSpy.mockReturnValue(scenarioUpdateHookValuesDefault)
    const component = create(getComponent())
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it("has correct structure", async () => {
    scenarioUpdateStateSpy.mockReturnValue(scenarioUpdateHookValuesDefault)
    const tree = shallow(getComponent())

    expect(tree.find(Columns)).toHaveLength(2)
    expect(tree.find(Column)).toHaveLength(6)
    expect(tree.find(InlineEditableHeaderContainer)).toHaveLength(1)
    expect(tree.find(ReadonlyActionField)).toHaveLength(2)
    expect(tree.find(MetaEntryTags)).toHaveLength(1)
    expect(tree.find(MetaEmail)).toHaveLength(1)
    expect(tree.find(MetaContributors)).toHaveLength(1)
  })
  it("has correct structure with edit date modal", async () => {
    scenarioUpdateStateSpy.mockReturnValue(scenarioUpdateHookValuesDefault)
    const tree = shallow(getComponent({isEditDateModalVisible: true}))

    expect(tree.find(Columns)).toHaveLength(2)
    expect(tree.find(Column)).toHaveLength(6)
    expect(tree.find(InlineEditableHeaderContainer)).toHaveLength(1)
    expect(tree.find(ReadonlyActionField)).toHaveLength(2)
    expect(tree.find(MetaEntryTags)).toHaveLength(1)
    expect(tree.find(MetaEmail)).toHaveLength(1)
    expect(tree.find(MetaContributors)).toHaveLength(1)
    expect(tree.find(EditScenarioDateModalContainer)).toHaveLength(1)
    expect(tree.find(EditDurationModalContainer)).toHaveLength(0)
    expect(tree.find(InviteScenarioContributorsModalContainer)).toHaveLength(0)
  })
  it("has correct structure with invite modal", async () => {
    scenarioUpdateStateSpy.mockReturnValue(scenarioUpdateHookValuesDefault)
    const tree = shallow(getComponent({isInviteContributorsModalVisible: true}))

    expect(tree.find(Columns)).toHaveLength(2)
    expect(tree.find(Column)).toHaveLength(6)
    expect(tree.find(InlineEditableHeaderContainer)).toHaveLength(1)
    expect(tree.find(ReadonlyActionField)).toHaveLength(2)
    expect(tree.find(MetaEntryTags)).toHaveLength(1)
    expect(tree.find(MetaEmail)).toHaveLength(1)
    expect(tree.find(MetaContributors)).toHaveLength(1)
    expect(tree.find(EditScenarioDateModalContainer)).toHaveLength(0)
    expect(tree.find(EditDurationModalContainer)).toHaveLength(0)
    expect(tree.find(InviteScenarioContributorsModalContainer)).toHaveLength(1)
  })
  it("has correct structure with edit duration", async () => {
    scenarioUpdateStateSpy.mockReturnValue(scenarioUpdateHookValuesDefault)
    const tree = shallow(getComponent({isEditDurationModalVisible: true}))

    expect(tree.find(Columns)).toHaveLength(2)
    expect(tree.find(Column)).toHaveLength(6)
    expect(tree.find(InlineEditableHeaderContainer)).toHaveLength(1)
    expect(tree.find(ReadonlyActionField)).toHaveLength(2)
    expect(tree.find(MetaEntryTags)).toHaveLength(1)
    expect(tree.find(MetaEmail)).toHaveLength(1)
    expect(tree.find(MetaContributors)).toHaveLength(1)
    expect(tree.find(EditScenarioDateModalContainer)).toHaveLength(0)
    expect(tree.find(EditDurationModalContainer)).toHaveLength(1)
    expect(tree.find(InviteScenarioContributorsModalContainer)).toHaveLength(0)
  })

  it("renders correctly without time", async () => {
    scenarioUpdateStateSpy.mockReturnValue(scenarioUpdateHookValuesDefault)
    const tree = shallow(getComponent({scenario: {...scenariosMock[0], maxDurationInSeconds: null}}))

    const html = tree.html()
    expect(html).toContain("placeholder__no_entry")
    expect(html).not.toContain("null unit__seconds_short")
  })

  it("handles update", async () => {
    const mockUpdateScenario = jest.fn()

    scenarioUpdateStateSpy.mockReturnValue(scenarioUpdateHookValuesDefault)
    const tree = shallow(getComponent({updateScenario: mockUpdateScenario}))

    const inlineEditableHeaderContainers = tree.find(InlineEditableHeaderContainer)
    inlineEditableHeaderContainers.at(0).props().onConfirm("update")
    expect(mockUpdateScenario).toHaveBeenCalledWith(scenarioId, {...defaultUpdateData, name: "update"})

    const overlayEditField = tree.find(OverlayEditField)
    overlayEditField.props().onUpdate({description: "update"})
    expect(mockUpdateScenario).toHaveBeenCalledWith(scenarioId, {...defaultUpdateData, description: "update"})
  })
})
