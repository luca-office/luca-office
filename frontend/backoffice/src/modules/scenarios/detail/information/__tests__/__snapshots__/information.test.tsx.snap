// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`information renders correctly 1`] = `
<div
  className="card-content css-1vyr92j-CardContent"
>
  <div
    className="css-1ucjwqm-Columns"
  >
    <div
      className="css-me9x6r-Column"
    >
      <div
        className="css-lz4b3k-Label"
      >
        title
        <div
          className="css-nhflpf-Icon"
          dangerouslySetInnerHTML={
            {
              "__html": {},
            }
          }
        />
      </div>
      <div
        className="css-nvhwxh-InlineEditableHeader"
      >
        <div>
          <form
            className="css-1c5ej61-InlineEditableHeader"
            onBlur={[Function]}
            onSubmit={[Function]}
          >
            <div
              className="css-pl4s5v-InlineEditableHeader"
              onClick={[Function]}
            >
              <div
                className="css-1iwo50k"
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
              >
                <div
                  className="css-nrul0i"
                >
                  <input
                    className="css-q6vozw"
                    defaultValue="Scenario 1"
                    disabled={false}
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    readOnly={true}
                    type="text"
                  />
                </div>
              </div>
              <div
                className="hover-icon css-16vb3lw-Icon"
                dangerouslySetInnerHTML={
                  {
                    "__html": {},
                  }
                }
              />
            </div>
          </form>
        </div>
      </div>
    </div>
    <div
      className="css-1oiyj7r-Column"
    >
      <div
        className="css-cx62la-ReadonlyActionField"
        data-testid="readonly-action-field"
        onClick={[MockFunction]}
      >
        <div
          className="css-1bkess4-ReadonlyActionField"
        >
          <div
            className="css-47wnmy-ReadonlyActionField"
          >
            projects__detail_time_label
          </div>
        </div>
        <div
          className="content-wrapper css-b45n67-ReadonlyActionField"
        >
          <div
            className="content css-1hje2zc-ReadonlyActionField"
          >
            <div
              className="css-e8i3t7-Text"
            >
              <div
                className="css-1lrzij5-Icon"
                dangerouslySetInnerHTML={
                  {
                    "__html": {},
                  }
                }
              />
              90 unit__minutes_short
            </div>
          </div>
          <button
            className="readonly-action-button css-4nv627-Button"
            disabled={false}
            onClick={[Function]}
          >
            <div
              className="css-95j2hm-Button"
            >
              edit_button
            </div>
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    className="css-18ypg4c-Column"
  >
    <div
      className="css-lz4b3k-Label"
    >
      description
      <div
        className="css-nhflpf-Icon"
        dangerouslySetInnerHTML={
          {
            "__html": {},
          }
        }
      />
    </div>
    <div
      className="overlay-edit-field css-0"
    >
      <div
        className="css-1kjt8hf-ReadonlyActionField"
        data-testid="readonly-action-field"
        onClick={[Function]}
      >
        <div
          className="css-1bkess4-ReadonlyActionField"
        />
        <div
          className="content-wrapper css-376q3o-ReadonlyActionField"
        >
          <div
            className="content css-ci7wec-ReadonlyActionField"
          >
            <div
              className="css-1h2eymq-InlineEditableTextarea"
              onClick={[Function]}
            >
              <div
                className="textarea css-1n5bn12-InlineEditableTextarea"
                contentEditable={false}
                suppressContentEditableWarning={true}
              >
                description
              </div>
              <div
                className="hover-icon css-1pndqgp-Icon"
                dangerouslySetInnerHTML={
                  {
                    "__html": {},
                  }
                }
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="css-1x21b1w-Information"
  >
    <div
      className="overlay-edit-field css-0"
    >
      <div
        className="css-1v0z0k-ReadonlyActionField"
        data-testid="readonly-action-field"
        onClick={[Function]}
      >
        <div
          className="css-1bkess4-ReadonlyActionField"
        >
          <div
            className="css-47wnmy-ReadonlyActionField"
          >
            scenario_details__general_tags
          </div>
        </div>
        <div
          className="content-wrapper css-b45n67-ReadonlyActionField"
        >
          <div
            className="content css-1hje2zc-ReadonlyActionField"
          >
            <div
              className="css-1ws8uc4-Tag"
            >
              <div
                className="css-12ia3h2-Text"
              >
                placeholder__tags
              </div>
            </div>
          </div>
          <button
            className="readonly-action-button css-4nv627-Button"
            disabled={false}
            onClick={[Function]}
          >
            <div
              className="css-95j2hm-Button"
            >
              edit_button
            </div>
          </button>
        </div>
      </div>
    </div>
    <div
      className="css-uietcf-Columns"
    >
      <div
        className="css-1oiyj7r-Column"
      >
        <div
          className="css-1v0z0k-ReadonlyActionField"
          data-testid="readonly-action-field"
          onClick={[Function]}
        >
          <div
            className="css-1bkess4-ReadonlyActionField"
          >
            <div
              className="css-47wnmy-ReadonlyActionField"
            >
              scenario_details__general_email
            </div>
          </div>
          <div
            className="content-wrapper css-b45n67-ReadonlyActionField"
          >
            <div
              className="content css-1hje2zc-ReadonlyActionField"
            >
              <div
                className="css-1bkess4-ReadonlyActionField"
              >
                <div
                  className="css-l0jg45-Icon"
                  dangerouslySetInnerHTML={
                    {
                      "__html": {},
                    }
                  }
                />
                <div
                  className="css-odwty4-Text"
                >
                  scenario_details__unavailable
                </div>
              </div>
            </div>
            <button
              className="readonly-action-button css-4nv627-Button"
              disabled={false}
              onClick={[Function]}
            >
              <div
                className="css-95j2hm-Button"
              >
                create_button
              </div>
            </button>
          </div>
        </div>
      </div>
      <div
        className="css-1oiyj7r-Column"
      >
        <div
          className="css-1v0z0k-ReadonlyActionField"
          data-testid="readonly-action-field"
          onClick={[MockFunction]}
        >
          <div
            className="css-1bkess4-ReadonlyActionField"
          >
            <div
              className="css-47wnmy-ReadonlyActionField"
            >
              scenario_details__date
            </div>
          </div>
          <div
            className="content-wrapper css-b45n67-ReadonlyActionField"
          >
            <div
              className="content css-1hje2zc-ReadonlyActionField"
            >
              <div
                className="css-1bkess4-ReadonlyActionField"
              >
                <div
                  className="css-l0jg45-Icon"
                  dangerouslySetInnerHTML={
                    {
                      "__html": {},
                    }
                  }
                />
                <div
                  className="css-odwty4-Text"
                >
                  01.01.1970 (scenario_details__fictive_date)
                </div>
              </div>
            </div>
            <button
              className="readonly-action-button css-4nv627-Button"
              disabled={false}
              onClick={[Function]}
            >
              <div
                className="css-95j2hm-Button"
              >
                edit_button
              </div>
            </button>
          </div>
        </div>
      </div>
      <div
        className="css-1bbhnip-Column"
      >
        <div
          className="css-1v0z0k-ReadonlyActionField"
          data-testid="readonly-action-field"
          onClick={[MockFunction]}
        >
          <div
            className="css-1bkess4-ReadonlyActionField"
          >
            <div
              className="css-47wnmy-ReadonlyActionField"
            >
              scenario_details__general_contributors
            </div>
          </div>
          <div
            className="content-wrapper css-b45n67-ReadonlyActionField"
          >
            <div
              className="content css-1hje2zc-ReadonlyActionField"
            >
              <div
                className="css-1bkess4-ReadonlyActionField"
              >
                <div
                  className="css-l0jg45-Icon"
                  dangerouslySetInnerHTML={
                    {
                      "__html": {},
                    }
                  }
                />
                <div
                  className="css-odwty4-Text"
                >
                  0 scenario_details__general_contributors
                </div>
              </div>
            </div>
            <button
              className="readonly-action-button css-4nv627-Button"
              disabled={false}
              onClick={[Function]}
            >
              <div
                className="css-95j2hm-Button"
              >
                scenario_details__button_invite
              </div>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
